java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -35,41 +34,79 @@ import kotlinx.coroutines.withContext
     /* internal */ val stores = LruCacheMap<String, StoreRecord<Data, Op, T>>(
         50,
         livenessPredicate = { _, sr -> !sr.store.closed }
-    ) { _, sr -> closeStore(sr) }
+    ) { muxId, sr -> closeStore(muxId, sr) }
+
+    // Keeps track of all the stores a callbackId is registered to. Used when removing an
+    // observer (off method), as that requires unregistering to each store it is registered with.
+    private val callbackIdToMuxIdMap = mutableMapOf<Int, MutableSet<String>>()
+
+    fun on(callback: ProxyCallback<Data, Op, T>, callbackToken: Int? = null): Int {
+        // Direct Store Muxer must generate the callback token to ensure no duplicates.
+        check(callbackToken == null) { "token must be generated by Direct Store Muxer" }
+        synchronized(proxyManager) {
+            val callbackId = proxyManager.register(callback, callbackToken)
+            callbackIdToMuxIdMap[callbackId] = mutableSetOf()
+            return callbackId
+        }
+    }
+
+    fun off(callbackToken: Int) {
+        synchronized(proxyManager) {
+            for (muxId in callbackIdToMuxIdMap.getOrElse(callbackToken, { mutableSetOf() })) {
+                val (idSet, store) = checkNotNull(stores[muxId]) { "store not found" }
+                store.off(callbackToken)
+                idSet.remove(callbackToken)
+            }
+            callbackIdToMuxIdMap.remove(callbackToken)
+            proxyManager.unregister(callbackToken)
+        }
+    }
 
     /** Safely closes a [DirectStore] and cleans up its resources. */
-    private fun closeStore(storeRecord: StoreRecord<*, *, *>) {
+    private fun closeStore(muxId: String, storeRecord: StoreRecord<*, *, *>) {
         if (!storeRecord.store.closed) {
-            log.debug { "close the store(${storeRecord.id})" }
+            log.debug { "close the store($muxId)" }
 
             try {
+                for (id in storeRecord.idSet) {
+                    storeRecord.store.off(id)
+                    callbackIdToMuxIdMap[id]?.remove(muxId)
+                }
                 storeRecord.store.close()
             } catch (e: Exception) {
                 // TODO(b/160251910): Make logging detail more cleanly conditional.
-                log.debug(e) { "failed to close the store(${storeRecord.id})" }
+                log.debug(e) { "failed to close the store($muxId)" }
                 log.info { "failed to close the store" }
             }
         }
     }
 
     /**
-     * Gets data from the store corresponding to the given [referenceId].
+     * Gets data from the store corresponding to the given [muxId].
      */
-    suspend fun getLocalData(referenceId: String) = store(referenceId).store.getLocalData()
+    suspend fun getLocalData(muxId: String, id: Int): Data {
+        // registerToStore will only register to the store if the id is not already registered
+        registerToStore(id, muxId)
+        val (_, store) = store(muxId)
```

Comment:
ok looking at this again. What about if we combine registerToStore() and store(). and this can become `val store = getStore(id, muxId)`. I think it makes sense as if you want to interact with the store, you need to register with it as well, I don't think it we call store() without registerToStore() anywhere.

suspend fun getStore(muxId: String) = storeMutex.withLock {
    val (idSet, store) = stores.getOrPut(muxId) {
        setupStore(muxId)
    }
    if (!idSet.contains(id)) {
        store.on(
            ProxyCallback { message ->
               proxyManager.getCallback(id)?.invoke(message.withMuxId(muxId))
            },
            id
        )
        idSet.add(id)
        callbackIdToMuxIdMap.getOrPut(id, { mutableSetOf() }).add(muxId)
    }
   store
}

WDYT? probably I am over-thinking this... just do whatever you prefer.

Reply:
This looks good. I've made the id parameter optional. Some tests needs the store record without having an ID to register to the store with.